### PR TITLE
[alpha.webkit.UncountedCallArgsChecker] Allow protector functions in Objective-C++

### PR DESCRIPTION
**Pull request by rniwa**
This PR fixes the bug that WebKit checkers didn't recognize the return value of an Objective-C++ selector which returns Ref or RefPtr to be safe.

### CODE REVIEW

File: clang/lib/StaticAnalyzer/Checkers/WebKit/PtrTypesSemantics.cpp [hunk 1]
```diff
@@ -143,6 +143,16 @@ bool isReturnValueRefCounted(const clang::FunctionDecl *F) {
   return false;
 }
 
+std::optional<bool> isUncounted(const clang::QualType T) {
```
Comment (haoNoQ) [inline clang/lib/StaticAnalyzer/Checkers/WebKit/PtrTypesSemantics.cpp:146]: `clang::` is redundant because you're in `using namespace clang`.

Comment (rniwa) [reply clang/lib/StaticAnalyzer/Checkers/WebKit/PtrTypesSemantics.cpp]: Fixed!

File: clang/lib/StaticAnalyzer/Checkers/WebKit/PtrTypesSemantics.cpp [hunk 2]
```diff
@@ -143,6 +143,16 @@ bool isReturnValueRefCounted(const clang::FunctionDecl *F) {
...
+  if (auto *Subst = dyn_cast<SubstTemplateTypeParmType>(T)) {
+    if (auto *Decl = Subst->getAssociatedDecl()) {
+      if (isRefType(safeGetName(Decl)))
+        return false;
+    }
+  }
+  return isUncounted(T->getAsCXXRecordDecl());
+}
+
 std::optional<bool> isUncounted(const CXXRecordDecl* Class)
```
Comment (haoNoQ) [inline clang/lib/StaticAnalyzer/Checkers/WebKit/PtrTypesSemantics.cpp:156]: Should we force every checker to use the new function now that we know about this cornercase?

**Review by haoNoQ — APPROVED** [review]
Aha ok LGTM!

File: clang/lib/StaticAnalyzer/Checkers/WebKit/PtrTypesSemantics.cpp (hunk 2 above)
Comment (rniwa) [reply clang/lib/StaticAnalyzer/Checkers/WebKit/PtrTypesSemantics.cpp:156]: So the only other place where this function is called is in `isUncountedPtr` but that function needs to evaluate whether the pointee is ref counted or not so I don't think we can use the new function.

**Comment by rniwa** [comment]
Thanks for the review!

**Pull request merged**